src/hotspot/share/c1/c1_Optimizer.cpp

```diff
@@ -882,7 +884,8 @@ void NullCheckEliminator::iterate_one(BlockBegin* block) {
     // visiting instructions which are references in other blocks or
     // visiting instructions more than once.
     mark_visitable(instr);
-    if (instr->is_pinned() || instr->can_trap() || (instr->as_NullCheck() != nullptr)) {
+    if (instr->is_pinned() || instr->can_trap() || (instr->as_NullCheck() != nullptr)
+        || (instr->as_Constant() != nullptr && instr->as_Constant()->type()->is_object())) {
```

Comment:
Is this just `instr->as_ObjectConstant() != nullptr`?

Reply:
Do you mean `insr->type()->as_ObjectConstant() != nullptr`?
But we should include other Constants (e.g. `ArrayConstant`, `InstanceConstant`), and those classes don't implement `as_ObjectConstant`

Reply:
Ah, OK then. Yes, I thought ObjectConstant includes ArrayConstant and InstanceConstant.